test/UniversalRouter.crossVersion.t.sol

```diff
@@ -0,0 +1,402 @@
+        );
+        (bool success,) = deployer.call(abi.encodeWithSignature("setFactoryAddress(address)", address(v3Factory)));
+        require(success, "setFactoryAddress failed");
+        v3Nfpm = IV3NonfungiblePositionManager(
+            createContractThroughBytecode(
+                _getNfpmBytecodePath(), toBytes32(deployer), toBytes32(address(v3Factory)), toBytes32(address(weth)), 0
+            )
+        );
+
+        ///////////////////////////////////
+        //////////// v4 setup /////////////
+        ///////////////////////////////////
+        vault = IVault(new Vault());
+        binPoolManager = new BinPoolManager(vault);
+        clPoolManager = new CLPoolManager(vault);
+        vault.registerApp(address(binPoolManager));
+        vault.registerApp(address(clPoolManager));
+
+        binPositionManager = new BinPositionManager(vault, binPoolManager, permit2);
+        clPositionManager = new CLPositionManager(vault, clPoolManager, permit2, 100_000);
+
+        ///////////////////////////////////
+        //////////// Router setup /////////////
+        ///////////////////////////////////
+        RouterParameters memory params = RouterParameters({
+            permit2: address(permit2),
+            weth9: address(weth),
+            v2Factory: address(0),
+            v3Factory: address(v3Factory),
+            v3Deployer: deployer,
+            v2InitCodeHash: bytes32(0),
+            v3InitCodeHash: bytes32(0x6ce8eb472fa82df5469c6ab6d485f17c3ad13c8cd7af59b3d4a8026c5ce0f7e2),
+            stableFactory: address(0),
+            stableInfo: address(0),
+            v4Vault: address(vault),
+            v4ClPoolManager: address(clPoolManager),
+            v4BinPoolManager: address(binPoolManager),
+            v3NFTPositionManager: address(v3Nfpm),
+            v4ClPositionManager: address(clPositionManager),
+            v4BinPositionManager: address(binPositionManager)
+        });
+        router = new UniversalRouter(params);
+        _approvePermit2ForCurrency(address(this), currency0, address(router), permit2);
+        _approvePermit2ForCurrency(address(this), currency1, address(router), permit2);
+        _approvePermit2ForCurrency(address(this), Currency.wrap(address(weth)), address(router), permit2);
+
+        ///////////////////////////////////
+        //////////// Add Liquidity /////////////
+        ///////////////////////////////////
+
+        // add liquidity to v3 usdt-weth pool
+        _mintV3Liquidity(address(usdt), address(weth), liquidityProvider);
+
+        // add liquidity to v4 usdc-eth cl-pool
+        clPoolKeyWithETH = _mintV4CLLiquidity(address(usdc), address(0), liquidityProvider);
+        // add liquidity to v4 usdc-weth cl-pool
+        clPoolKeyWithWrappedETH = _mintV4CLLiquidity(address(usdc), address(weth), liquidityProvider);
+    }
+
+    /// @dev USDT -> v3 (WETH - USDT) -> v4 (ETH - USDC) -> USDC
```

Comment:
able to make it easier for others by adding this comment:

`/// @dev case0: swap with a v4 native (eth-usdc) pool` 

and for case 1:

`/// @dev case1: swap with a v4 weth-usdc pool` 


took a while to see the differences lol